File: src/services/odsaCompanion.ts
/**
 * ODSA Companion Device Handler (ap2006)
 *
 * Handles eSIM lifecycle operations for companion devices.
 */

import type { ApplicationConfig } from '../protocol/responseTypes.js';
import { EntitlementStatus, SubscriptionResult } from '../protocol/statusCodes.js';
import { getActivationCode } from './mockSmdp.js';
import type { OdsaConfigData, OdsaContext } from './odsaCommon.js';
import { buildOdsaBaseConfig } from './odsaCommon.js';

export function buildCompanionConfig(
  status: number,
  provStatus: number,
  tcStatus: number,
  configData?: unknown,
  odsaContext?: OdsaContext,
): ApplicationConfig {
  const data = (configData ?? {}) as OdsaConfigData;
  const operation = odsaContext?.operation;

  if (!operation) {
    // No operation — basic entitlement status only
    return {
      appId: 'ap2006',
      entitlementStatus: status,
      provStatus,
      tcStatus,
    };
  }

  switch (operation) {
    case 'CheckEligibility':
      return handleCheckEligibility(status, provStatus, tcStatus, data);
    case 'ManageSubscription':
      return handleManageSubscription(status, provStatus, tcStatus, data);
    case 'ManageService':
      return handleManageService(status, provStatus, tcStatus, data);
    case 'AcquireConfiguration':
      return handleAcquireConfiguration(status, provStatus, tcStatus, data);
    case 'AcquireTemporaryToken':
      return buildOdsaBaseConfig('ap2006', status, provStatus, tcStatus, SubscriptionResult.DONE);
    default:
      return {
        appId: 'ap2006',
        entitlementStatus: status,
        provStatus,
        tcStatus,
      };
  }
}

function handleCheckEligibility(
  status: number, provStatus: number, tcStatus: number, data: OdsaConfigData,
): ApplicationConfig {
  const isEligible = data.subscriptionState === 'eligible' || data.subscriptionState === 'active';
  const entitlementStatus = isEligible ? EntitlementStatus.ENABLED : EntitlementStatus.DISABLED;
  return buildOdsaBaseConfig('ap2006', entitlementStatus, provStatus, tcStatus, SubscriptionResult.DONE);
}

function handleManageSubscription(
  status: number, provStatus: number, tcStatus: number, data: OdsaConfigData,
): ApplicationConfig {
  // Eligible + serviceFlowUrl → redirect to web portal
  if (data.subscriptionState === 'eligible' && data.serviceFlowUrl) {
    const config = buildOdsaBaseConfig(
      'ap2006', status, provStatus, tcStatus, SubscriptionResult.CONTINUE_TO_WS,
      { ServiceFlow_URL: data.serviceFlowUrl },
    );
    if (data.serviceFlowUserData) {
      config.extraParams!.ServiceFlow_UserData = data.serviceFlowUserData;
    }
    return config;
  }

  // Active + smdpAddress → download eSIM profile
  if (data.subscriptionState === 'active' && data.smdpAddress) {
    const profileKey = data.profileType ?? 'companion';
    const activation = getActivationCode(profileKey);
    return buildOdsaBaseConfig(
      'ap2006', status, provStatus, tcStatus, SubscriptionResult.DOWNLOAD_PROFILE,
      {
        'SMDP+Address': data.smdpAddress,
        'SMDP+ActivationCode': activation.activationCode,
        ProfileICCID: activation.iccid,
      },
    );
  }

  // Default
  return buildOdsaBaseConfig('ap2006', status, provStatus, tcStatus, SubscriptionResult.DONE);
}

function handleManageService(
  status: number, provStatus: number, tcStatus: number, _data: OdsaConfigData,
): ApplicationConfig {
  return buildOdsaBaseConfig('ap2006', status, provStatus, tcStatus, SubscriptionResult.DONE);
}

function handleAcquireConfiguration(
  status: number, provStatus: number, tcStatus: number, data: OdsaConfigData,
): ApplicationConfig {
  const extra: Record<string, string> = {};
  if (data.smdpAddress) {
    extra['SMDP+Address'] = data.smdpAddress;
  }
  if (data.profileIccid) {
    extra.ProfileICCID = data.profileIccid;
  }
  return buildOdsaBaseConfig('ap2006', status, provStatus, tcStatus, SubscriptionResult.DONE, extra);
}

File: internal/services/odsa_companion.go
package services

import (
	"github.com/hkassaei/poc-entitlements-as-a-service-go/internal/protocol"
)

// BuildCompanionConfig builds the ApplicationConfig for ODSA Companion (ap2006).
func BuildCompanionConfig(status, provStatus, tcStatus int, configData *OdsaConfigData, ctx *OdsaContext) *protocol.ApplicationConfig {
	if configData == nil {
		configData = &OdsaConfigData{}
	}

	operation := ""
	if ctx != nil {
		operation = ctx.Operation
	}

	if operation == "" {
		return &protocol.ApplicationConfig{
			AppID:             "ap2006",
			EntitlementStatus: status,
			ProvStatus:        protocol.IntPtr(provStatus),
			TcStatus:          protocol.IntPtr(tcStatus),
		}
	}

	switch operation {
	case protocol.OpCheckEligibility:
		return companionCheckEligibility(provStatus, tcStatus, configData)
	case protocol.OpManageSubscription:
		return companionManageSubscription(status, provStatus, tcStatus, configData)
	case protocol.OpManageService:
		return BuildOdsaBaseConfig("ap2006", status, provStatus, tcStatus, protocol.SubscriptionResultDone, nil)
	case protocol.OpAcquireConfiguration:
		return companionAcquireConfiguration(status, provStatus, tcStatus, configData)
	case protocol.OpAcquireTemporaryToken:
		return BuildOdsaBaseConfig("ap2006", status, provStatus, tcStatus, protocol.SubscriptionResultDone, nil)
	default:
		return &protocol.ApplicationConfig{
			AppID:             "ap2006",
			EntitlementStatus: status,
			ProvStatus:        protocol.IntPtr(provStatus),
			TcStatus:          protocol.IntPtr(tcStatus),
		}
	}
}

func companionCheckEligibility(provStatus, tcStatus int, data *OdsaConfigData) *protocol.ApplicationConfig {
	status := protocol.EntitlementStatusDisabled
	if data.SubscriptionState == "eligible" || data.SubscriptionState == "active" {
		status = protocol.EntitlementStatusEnabled
	}
	return BuildOdsaBaseConfig("ap2006", status, provStatus, tcStatus, protocol.SubscriptionResultDone, nil)
}

func companionManageSubscription(status, provStatus, tcStatus int, data *OdsaConfigData) *protocol.ApplicationConfig {
	if data.SubscriptionState == "eligible" && data.ServiceFlowURL != "" {
		extra := map[string]string{"ServiceFlow_URL": data.ServiceFlowURL}
		if data.ServiceFlowUserData != "" {
			extra["ServiceFlow_UserData"] = data.ServiceFlowUserData
		}
		return BuildOdsaBaseConfig("ap2006", status, provStatus, tcStatus, protocol.SubscriptionResultContinueToWS, extra)
	}

	if data.SubscriptionState == "active" && data.SMDPAddress != "" {
		profileKey := data.ProfileType
		if profileKey == "" {
			profileKey = "companion"
		}
		activation := GetActivationCode(profileKey)
		return BuildOdsaBaseConfig("ap2006", status, provStatus, tcStatus, protocol.SubscriptionResultDownloadProfile, map[string]string{
			"SMDP+Address":        data.SMDPAddress,
			"SMDP+ActivationCode": activation.ActivationCode,
			"ProfileICCID":        activation.ICCID,
		})
	}

	return BuildOdsaBaseConfig("ap2006", status, provStatus, tcStatus, protocol.SubscriptionResultDone, nil)
}

func companionAcquireConfiguration(status, provStatus, tcStatus int, data *OdsaConfigData) *protocol.ApplicationConfig {
	extra := map[string]string{}
	if data.SMDPAddress != "" {
		extra["SMDP+Address"] = data.SMDPAddress
	}
	if data.ProfileICCID != "" {
		extra["ProfileICCID"] = data.ProfileICCID
	}
	return BuildOdsaBaseConfig("ap2006", status, provStatus, tcStatus, protocol.SubscriptionResultDone, extra)
}
